### DISCUSSION TITLE
Pushing extra values from modbus to MQTT

### DISCUSSION THREAD

**Discussion by devdems** [dingo35/SmartEVSE-3.5#56] [General]
Hi @dingo35, based on this feature request https://github.com/serkri/SmartEVSE-3/issues/245

what would be the easiest way to get a value from Modbus on an address that is not defined in EMstruct to MQTT?

I would like to do something like this:

if (MQTTclient.connected()) {
        if (MainsMeter) {
            MQTTclient.publish(MQTTprefix + "/MainsCurrentL1", String(Irms[0]), false, 0);
            MQTTclient.publish(MQTTprefix + "/MainsCurrentL2", String(Irms[1]), false, 0);
            MQTTclient.publish(MQTTprefix + "/MainsCurrentL3", String(Irms[2]), false, 0);
            **MQTTclient.publish(MQTTprefix + "/MainsCurrentPower", String(----Get value from registry address 40226 from mains meter----), false, 0);**
        }

**Comment by dingo35** [comment]
Well that turned out the be a very difficult thing; I had a flexible solution for that, the tcp modbus bridge, but that disturbed the modbus traffic.

Turns out there is a timing problem: look at the modbusRequest loop, every request for every device, every address you do, has to be programmed in that loop by hand (not difficult, but not flexible), and all those request have to fit in the loop period of time.

Second problem is even bigger: the modbus library puts requests on the bus, but the response that comes back only contains the device address and the function nr. that was used to make the request (usually 3 or 4, but can also be 6 or 10).
So when the SmartEVSE receives the response, it has to find out which request this was a response to. IMHO this is a big drawback of the modbus protocol....

So there is no easy solution for this. Im still hoping for a modbus library that makes these problems go away, so we can enable the tcp modbus bridge again and everybody can get all the stuff off of the modbus that they think is useful....

**Comment by devdems** [comment]
I did just this yes. Tnx for your advice. It is a very ugly solution but it works. I will not be making a PR as it does not make sense for the wider audience.

If something will come up to make this feature more user-friendly it would be great. I'm closing this now.
